teps/0109-better-structured-provenance-retrieval-in-tekton-chains.md

```diff
@@ -0,0 +1,502 @@
+```
+{
+        "uri": "<URI>",
+        "digest": { /* DigestSet */ }
+ }
+```
+The other specific requirements of these fields can be found in [in-toto attestation spec](https://github.com/in-toto/attestation/tree/main/spec). Another thing to be aware is that For `digest`, it has a key for the digest type and has a value for that key. Git, Mercurial, Bazaar and Subversion have full VCS location support. CVS, Perforce are providing this VCS location as plain URLs. For VCS, the list of VCS that the community are looking into is: Git, Mercurial, Subversion, Bazaar, CVS, Perforce. Currently Chains only supports Git revisions for VCS and OCI images as signable targets for inputs.
+
+For the `outputs`, we are looking at the following support, as these are the artifact types supported in Artifact Registries. However, the list should definitely be expanded as this doc is spreading the discussions in the communities to see which should be prioritized. Currently, we have the following under considerations: Python, Maven, Go, NodeJS and OCI images. Note that OCI images are the ones we currently support.
+
+#### Concrete Use Cases
+All below examples will be generating in-toto attestation.
+ 1. Use Git commits as sources, and generates in-toto provenance for a TaskRun that builds an image.
+   
+    A Task example `without` structured results and params support can be like:
+    ``` yaml
+    apiVersion: tekton.dev/v1beta1
+    kind: Task
+    spec:
+      params:
+        - name: CHAINS-GIT_COMMIT
+          type: string
+          description: git commit sha
+        - name: CHAINS-GIT_URL
+          type: string
+          description: git commit url
+        - name: FOO_IMAGE
+          type: string
+        - description: artifact image
+      ...
+      results:
+        - name: IMAGE_DIGEST
+          description: digest of image
+        - name: IMAGE_URL
+          description: url of image
+    ```
+    A Task example `with` structured results support will be like:
+    ``` yaml
+    apiVersion: tekton.dev/v1beta1
+    kind: Task
+    spec:
+      ...
+      results:
+        - name: ARTIFACT_INPUTS
+          type: array
+          description: |
+            Stores result names that should be captured as signable targets as "materials" in in-toto provenances in Tekton Chains.
+        - name: ARTIFACT_OUTPUTs
+          type: array
+          description: |
+            Stores result names that should be captured as signable targets as "subjects" in in-toto provenances in Tekton Chains.
+        - name: git-vcs
+          type: object
+          description: |
+            The source distribution
+            * uri: resource uri of the artifact.
+            * digest: revision digest in form of algorithm:digest.
```

Comment:
How are we planning to handle DigestSets?

Reply:
Can you provide more context here?

Reply:
Right now we have uri as type `string`, but SLSA defines this as `map[string]string`, so this isn't compatible with the intoto spec.

Reply:
Ah I see your point now. If we want DigestSets to be supported in an organizational way, we need Pipeline Results to have nested objects to be supported. I added a DigestSets section in the doc

Reply:
Do we have a way of representing variable length outputs? e.g. how the `IMAGES` field works today.

I'm trying to think through how things where we may not know the exact # of artifacts produced like multiarch images can be supported. 🤔

Reply:
That's a super good point! I was investigating in-toto provenance, I think `subjects` and `materials` seem not compatible with multiarch images. `DigestSets` is a `map[string]string`, but the examples in the in-toto repo only shows examples that key is an algorithm and value is the digest itself. I filed https://github.com/in-toto/attestation/issues/105 for tracking that

Reply:
SLSA team suggested to list them separately and add unique names for those or just add the digest for the image manifest.